modules/mbedtls_curl/http_client_curl.cpp

```diff
@@ -0,0 +1,440 @@
+			curl_easy_cleanup(msg->easy_handle);
+			in_flight = false;
+		}
+	}
+	return OK;
+}
+
+RingBuffer<uint8_t> *HTTPClientCurl::_init_upload(CURL *p_chandle, Method p_method, uint8_t *p_body, int p_body_size) {
+	RingBuffer<uint8_t> *b = memnew(RingBuffer<uint8_t>);
+	b->resize(p_body_size);
+	b->write(p_body, p_body_size);
+
+	// Special cases for POST and PUT to configure uploads.
+	switch (p_method) {
+		case METHOD_GET:
+			break;
+		case METHOD_HEAD:
+			break;
+		case METHOD_POST:
+			curl_easy_setopt(p_chandle, CURLOPT_POST, 1L);
+			curl_easy_setopt(p_chandle, CURLOPT_POSTFIELDSIZE_LARGE, (curl_off_t)p_body_size);
+			break;
+		case METHOD_PUT:
+			curl_easy_setopt(p_chandle, CURLOPT_UPLOAD, 1L);
+			curl_easy_setopt(p_chandle, CURLOPT_INFILESIZE_LARGE, (curl_off_t)p_body_size);
+			break;
+		case METHOD_DELETE:
+			break;
+		case METHOD_OPTIONS:
+			break;
+		case METHOD_TRACE:
+			break;
+		case METHOD_CONNECT:
+			break;
+		case METHOD_PATCH:
+			break;
+		case METHOD_MAX:
+			break;
+	}
+
+	// Somewhat counter intuitively, the read function is actually used by libcurl to send data,
+	// while the write function (see below) is used by libcurl to write response data to storage
+	// (or in our case, memory).
+	curl_easy_setopt(p_chandle, CURLOPT_READFUNCTION, _read_callback);
+	curl_easy_setopt(p_chandle, CURLOPT_READDATA, b);
+	return b;
+}
+
+RequestContext *HTTPClientCurl::_create_request_context() {
+	RequestContext *ctx = memnew(RequestContext);
+	ctx->response_headers = &response_headers;
+	ctx->body_size = &body_size;
+	ctx->status = &status;
+	ctx->response_chunks = &response_chunks;
+	ctx->has_response = &response_available;
+	ctx->chunked = &chunked;
+	ctx->keep_alive = &keep_alive;
+	return ctx;
+}
+
+Error HTTPClientCurl::_init_dns(CURL *p_chandle, IPAddress p_addr) {
+	// TODO: Support resolving multiple addresses.
+	curl_slist *h = _ip_addr_to_slist(p_addr);
+	CURLcode return_code = curl_easy_setopt(p_chandle, CURLOPT_RESOLVE, h);
+	if (return_code != CURLE_OK) {
+		ERR_PRINT("failed to initialize dns resolver: " + String::num_int64(return_code));
+		return FAILED;
+	}
+	return OK;
+}
+
+Error HTTPClientCurl::_init_request_headers(CURL *p_chandler, Vector<String> p_headers, RequestContext *p_ctx) {
+	for (int i = 0; i < p_headers.size(); i++) {
+		p_ctx->header_list = curl_slist_append(p_ctx->header_list, p_headers[i].ascii().get_data());
+	}
+	if (p_ctx->header_list) {
+		CURLcode return_code = curl_easy_setopt(p_chandler, CURLOPT_HTTPHEADER, p_ctx->header_list);
+		if (return_code != CURLE_OK) {
+			ERR_PRINT("failed to set request headers: " + String::num_uint64(return_code));
+			return FAILED;
+		}
+	}
+	return OK;
+}
+
+Error HTTPClientCurl::_request(IPAddress p_addr, bool p_init_dns) {
+	String h = p_addr;
+	if (h.find(":") != -1) {
+		h = "[" + h + "]";
+	}
+	CURL *eh = curl_easy_init();
+	curl_easy_setopt(eh, CURLOPT_URL, (scheme + h + ":" + String::num_int64(port) + url).ascii().get_data());
```

Comment:
Here, we don't want to set the ip address as the `CURLOPT_URL` but the actual hostname (this is what will be used for the `Host: hostname:port` header).

Reply:
Okay this is fixed. And it also makes `http://www.example.com` finally work as well!